DM-40395: Allow datastore record inserts to use ensure or replace

timj opened the pull request:
This doesn't solve the real problem of the files being overwritten without us knowing whether the datastore records are allowed to be modified but it does give the option.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

timj committed:
Formally deprecate the idGenerationMode parameter for Butler.ingest

This should have been done when we removed the support for ingesting
unresolved refs.

timj committed:
Add ensure and/or replace variants for opaque and bridge interface

timj committed:
Allow inserts of records to use ensure or replace rather than insert

timj committed:
Special case ingest to use "replace" for record insertion in some situations

Until we have a coherent way of handling record clashes, for ingest
assume that if UUIDv5 IDs are being used and the file is not
going to be transferred into the datastore, then replacing
existing records is allowable.

timj committed:
Test that repeated ingest with UUIDv5 + direct works

andy-slac commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -365,10 +365,20 @@ def _delete_artifact(self, location: Location) -> None:
 
-    def addStoredItemInfo(self, refs: Iterable[DatasetRef], infos: Iterable[StoredFileInfo]) -> None:
+    def addStoredItemInfo(
+        self, refs: Iterable[DatasetRef], infos: Iterable[StoredFileInfo], insert_mode: str = "insert"
```
Maybe use `Literal["insert", ...]` instead of `str`?

timj replied:
Should I use an enum? This is an internal interface so maybe that is better.

andy-slac replied:
👍 Enum is probably more natural here.

andy-slac commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -1036,7 +1047,16 @@ def _finishIngest(
-        self._register_datasets(refsAndInfos)
+            for ref in dataset.refs:
+                if ref.id.version != 5:
+                    uses_uuid_v5 = False
```
What happens if there is a mix of v4/v5 UUIDs? Should we split them into two groups? If we don't care then maybe add `break` after this assignment?

timj replied:
Hmm. Maybe I should split them into groups. I'm not expecting it to happen but you are right that it's the consistent thing to do with the current brokenness.

timj replied:
I now split them up.

andy-slac commented on python/lsst/daf/butler/datastores/fileDatastore.py:
```diff
@@ -2729,7 +2749,10 @@ def transfer_from(
-        self._register_datasets(artifacts)
+        # We are overwriting previous datasets that may have already
+        # existed. We therefore should ensure that we force the
+        # datastore records to agree.
```
Could it happen that old and new datasets have different set of components?

timj replied:
For ingest we do not do disassembly so replace should always be fine. For arbitrary transfers in this method I suppose it is possible that someone ran the graph in disassemble mode, transferred it, then re-ran the graph in standard mode and tried to transfer it again. We don't really want that to happen from a provenance perspective and ideally transferring something that you've already got should result in it being skipped -- `registry._insertDatasets` doesn't help us here and race conditions in datastore if transfers are happening concurrently doesn't help us either. I think I should switch it back to "insert" for now and leave the real fix for the other ticket.

timj replied:
I've added an extra commentary but still use replace. We really need to sort this out properly.

andy-slac commented on tests/test_datastore.py:
```diff
@@ -1101,6 +1102,36 @@ def testChecksum(self) -> None:
+
+            datastore.ingest(FileDataset(path=path, refs=v4ref), transfer="direct")
+
+            # This will fail because the ref is not using UUIDv4.
```
```suggestion
            # This will fail because the ref is using UUIDv4.
```
?

andy-slac reviewed (approved):
Looks good, a couple of minor questions.

timj committed:
Use an enum for insert_mode rather than a string

timj committed:
Fix mypy error with pydantic v1

timj committed:
Allow an insert to include uuid v5 and v4 refs

timj committed:
In Datastore.transfer_from use REPLACE mode for incoming records

This is a questionable change and needs some further thought,
but it is more consistent with what is actually happening and
the chances of a disassembled composite and an assembled
composite clashing are low.
